scene hierarchy prototype

Medyan-Naser opened the pull request:
### Describe your changes
Create prototype of scene hierarchy display and control using the vtkGLTFImporter

### Issue ticket number and link if any
#2491 
### Checklist for finalizing the PR

- [ ] I have performed a [self-review](https://f3d.app/doc/dev/CODING_STYLE.html) of my code
- [ ] I have added [tests](https://f3d.app/doc/dev/TESTING.html) for new features and bugfixes
- [ ] I have added [documentation](https://f3d.app/) for new features
- [ ] If it is a modifying the libf3d API, I have updated bindings
- [ ] If it is a modifying the `.github/workflows/versions.json`, I have updated `timestamp`

### Continuous integration

Please write a comment to run CI, eg: `\ci fast`.
See [here](https://f3d.app/CONTRIBUTING.html#continuous-integration) for more info.

mwestphal commented on vtkext/public/module/F3DNodeInfo.h:
does this needs to be in public ?

Medyan-Naser replied:
moved to private

mwestphal replied:
still in public ?

mwestphal commented on library/public/window.h:
```diff
@@ -119,6 +123,11 @@ class F3D_EXPORT window
+  /**
+   * Set the scene hierarchy for display in UI
+   */
+  virtual void SetSceneHierarchy(const std::vector<NodeInfo>& hierarchy) = 0;
```
def should not be public

Medyan-Naser replied:
moved to private

mwestphal commented on vtkext/private/module/vtkF3DRenderer.h:
```diff
@@ -442,6 +445,11 @@ class vtkF3DRenderer : public vtkOpenGLRenderer
+  /**
+   * Set a callback to request a render safely from within the render loop
+   */
+  void SetRenderRequestCallback(std::function<void()> callback);
```
im quite sure these is already a way to ImGui to trigger a safe render

mwestphal reviewed (changes requested):
changes needed

mwestphal commented:
Hi @Medyan-Naser 

Still working on this ? :)

mwestphal commented:
Also please fix the conflict with master.

mwestphal reviewed (commented):
changes needed before a review.

Medyan-Naser commented on vtkext/private/module/vtkF3DMetaImporter.cxx:
```diff
@@ -670,3 +677,117 @@ vtkMTimeType vtkF3DMetaImporter::GetUpdateMTime()
+{
+  std::vector<NodeInfo> nodes;
+
+  // Get scene hierarchy for GLTF files (where names are stored)
```
The implementation still uses vtkF3DGLTFImporter internally (as it's currently the only importer with scene hierarchy support), but the API and comments are now generic so other importers can be added in the future.

However, for any file type, there will be a root Actor in which its visibility can be switched.

mwestphal commented on library/public/scene.h:
```diff
@@ -148,6 +150,13 @@ class F3D_EXPORT scene
+   * Returns a vector of NodeInfo structures representing the actors in the scene.
+   * This is used to populate the scene hierarchy UI widget.
+   */
+  virtual std::vector<NodeInfo> GetSceneHierarchyNodes() = 0;
```
I mean that there is no reason for this method to be in the public API of scene.h, it should ba an implementation only API, in scene_impl.h

mwestphal commented on vtkext/private/module/vtkF3DImguiActor.cxx:
```diff
@@ -15,11 +15,16 @@
 #include <vtkVersion.h>
+#include <vtkCallbackCommand.h>
+#include <vtkCommand.h>
+#include <vtkSmartPointer.h>
```
sort

mwestphal commented on vtkext/private/module/vtkF3DImguiActor.cxx:
```diff
@@ -298,8 +304,8 @@ vtkStandardNewMacro(vtkF3DImguiActor);
 vtkF3DImguiActor::vtkF3DImguiActor()
-  : Pimpl(new Internals())
 {
+  this->Pimpl = std::make_unique<Internals>();
```
I dont think this is needed ?

mwestphal commented on vtkext/private/module/vtkF3DImguiActor.cxx:
```diff
@@ -873,6 +1000,7 @@ void vtkF3DImguiActor::EndFrame(vtkOpenGLRenderWindow* renWin)
   this->Pimpl->RenderDrawData(renWin, ImGui::GetDrawData());
 }
 
+
```
not needed

mwestphal commented on vtkext/private/module/vtkF3DMetaImporter.h:
```diff
@@ -22,7 +23,10 @@
-#include <optional>
+#include "vtkF3DImporter.h"
+
+#include "F3DNodeInfo.h"
```
put these include before vtk includes

mwestphal commented on vtkext/private/module/vtkF3DRenderer.cxx:
```diff
@@ -2917,6 +2940,7 @@ void vtkF3DRenderer::ConfigureColoring()
             this->UseCellColoring);
         }
       }
+      
```
not needed

mwestphal commented on vtkext/private/module/vtkF3DRenderer.h:
```diff
@@ -486,6 +489,8 @@ class vtkF3DRenderer : public vtkOpenGLRenderer
    */
   void SetUIDeltaTime(double time);
 
+  bool ConsumeUIRenderRequest();
```
what is this needed for ?

mwestphal commented on vtkext/private/module/CMakeLists.txt:
```diff
@@ -138,6 +138,10 @@ vtk_module_add_module(f3d::vtkextPrivate
 
+vtk_module_include(f3d::vtkextPrivate
+  PUBLIC
+    "${F3D_SOURCE_DIR}/library/private")
```
still wrong, lets discuss design on discord when you can :)

Medyan-Naser commented:
Why two renders are needed for ImGui interactions:

ImGui uses an immediate-mode rendering paradigm where input processing and state changes occur during the render pass, not before it. When a user clicks a checkbox in the scene hierarchy:

Mouse click event triggers RenderUI()
During Render(), ImGui processes the frame and detects the checkbox was clicked
Visibility is toggled via SetVisibility() and Modified()
But this happens mid-render, too late to affect the current frame
This differs from F3D hotkeys (like 'H' for help) which are processed before render starts, so the state change is visible immediately.

Simplified solution:

Instead of using a dedicated RequestRender()/ConsumeRenderRequest() callback mechanism, I simplified the logic to just do two renders on mouse button release.
Mouse button releases use fullRender=true since that's when ImGui click actions complete. This is simpler than tracking render requests with custom flags while still handling the ImGui limitation.

mwestphal commented on library/private/window_impl.h:
```diff
@@ -15,6 +15,7 @@
 #include "log.h"
 #include "window.h"
 
+#include "F3DNodeInfo.h"
```
still wrongly placed

mwestphal commented on vtkext/private/module/vtkF3DRenderer.h:
```diff
@@ -120,6 +118,7 @@ class vtkF3DRenderer : public vtkOpenGLRenderer
   void SetFilenameInfo(const std::string& info);
   void SetDropZoneInfo(const std::string& info);
   void SetDropZoneBinds(const std::vector<std::pair<std::string, std::string>>& dropZoneBinds);
+  void SetHierarchy(const std::vector<NodeInfo>& hierarchy);
```
SetSceneHierarchy

mwestphal commented on library/private/window_impl.h:
```diff
@@ -61,6 +62,13 @@ class window_impl : public window
+  /**
+   * Implementation only API.
+   * Set the scene hierarchy for display in UI.
+   * This is called internally by the scene when content changes.
```
```suggestion
   * This is called by the scene when content changes.
```

mwestphal commented on vtkext/private/module/F3DNodeInfo.h:
```diff
@@ -0,0 +1,25 @@
+
+class vtkProp;
+
+struct NodeInfo
```
```suggestion
struct F3DNodeInfo
```

to avoid collisions

mwestphal commented on vtkext/private/module/vtkF3DImguiActor.cxx:
```diff
@@ -372,13 +378,134 @@ void vtkF3DImguiActor::Initialize(vtkOpenGLRenderWindow* renWin)
+    return;
+  }
+
+  if (node->prop)
```
can this be nullptr ?

Medyan-Naser replied:
Yes, node->prop can be nullptr. Nodes without associated props (like group/container nodes) are rendered as plain text labels without checkboxes

mwestphal commented on vtkext/private/module/vtkF3DImguiActor.cxx:
```diff
@@ -372,13 +378,134 @@ void vtkF3DImguiActor::Initialize(vtkOpenGLRenderWindow* renWin)
+
+  if (node->prop)
+  {
+    ImGui::PushID((void*)node->prop);
```
static_cast

mwestphal commented on vtkext/private/module/vtkF3DImguiActor.cxx:
```diff
@@ -372,13 +378,134 @@ void vtkF3DImguiActor::Initialize(vtkOpenGLRenderWindow* renWin)
+  float totalWidth = maxWidth + checkboxWidth + 2.0f * ImGui::GetStyle().WindowPadding.x + 30.0f;
+
+  // Clamp to reasonable bounds
+  return std::max(200.0f, std::min(totalWidth, 800.0f));
```
use constexpr for such values

mwestphal commented on vtkext/private/module/vtkF3DMetaImporter.h:
```diff
@@ -166,6 +167,20 @@ class vtkF3DMetaImporter : public vtkF3DImporter
+  * Returns the scene hierarchy if any of the importers supports it.
+  * Returns nullptr if no importer with hierarchy support was added.
+  */
+  vtkDataAssembly* GetSceneHierarchy();
```
no reason for this not to be `GetSceneHierarchy () override`

mwestphal commented on vtkext/private/module/vtkF3DUIActor.h:
```diff
@@ -8,10 +8,17 @@
+#include <string>
+#include <vector>
+
+#include "F3DNodeInfo.h"
```
before vtk includes

Medyan-Naser commented on vtkext/private/module/vtkF3DImguiActor.cxx:
```diff
@@ -372,13 +378,134 @@ void vtkF3DImguiActor::Initialize(vtkOpenGLRenderWindow* renWin)
+  }
+
+  const ImGuiViewport* viewport = ImGui::GetMainViewport();
+  if (!viewport)
```
Replaced nullptr check with assert

mwestphal commented on vtkext/private/module/vtkF3DImguiActor.cxx:
```diff
@@ -372,13 +378,134 @@ void vtkF3DImguiActor::Initialize(vtkOpenGLRenderWindow* renWin)
+  {
+    const std::string& displayText = node.displayName.empty() ? node.name : node.displayName;
+    ImVec2 textSize = ImGui::CalcTextSize(displayText.c_str());
+    float nodeWidth = textSize.x + depth * 10.0f; // 10px indent per level
```
lets make it a constexpr

Medyan-Naser commented:
I added test cases to verify the actor hierarchy can be shown for file types glb, gltf, and vtp.
I also added an interaction test that uses a file with multiple actors and hide one of them.

mwestphal commented on vtkext/private/module/vtkF3DRenderer.cxx:
```diff
@@ -1865,6 +1877,16 @@ void vtkF3DRenderer::ConfigureMetaData()
+  if (this->SceneHierarchyVisible != show)
+  {
+    this->SceneHierarchyVisible = show;
+    this->UIActor->SetSceneHierarchyVisibility(show);
```
this->CheatSheetConfigured = false;

mwestphal commented on TestInteractionSceneHierarchy.log:
please cleanup that log, you can look at other logs for help.
